### PR TITLE
Better support and testing for empty string_array

### PR DESCRIPTION
**Pull request by cottsay**
Specifically test for `string_array`s of size 0, and relax the null check in `rcutils_string_array_cmp` when it won't be used.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10737/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6137/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8743/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10627/)

### CODE REVIEW

File: src/string_array.c [hunk 1]
```diff
@@ -98,10 +98,6 @@ rcutils_string_array_cmp(
     lhs, "lhs string array is null", return RCUTILS_RET_INVALID_ARGUMENT);
   RCUTILS_CHECK_FOR_NULL_WITH_MSG(
     rhs, "rhs string array is null", return RCUTILS_RET_INVALID_ARGUMENT);
-  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
-    lhs->data, "lhs->data is null", return RCUTILS_RET_INVALID_ARGUMENT);
-  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
-    rhs->data, "rhs->data is null", return RCUTILS_RET_INVALID_ARGUMENT);
```
Comment (dirk-thomas) [inline src/string_array.c:104]: Without these checks one of the data pointers can be `NULL` which will be accessed below unconditionally:
* https://github.com/ros2/rcutils/pull/246/files#diff-c0cb908ca9bf786997651b24b643070cR118
* https://github.com/ros2/rcutils/pull/246/files#diff-c0cb908ca9bf786997651b24b643070cR120

Comment (cottsay) [reply src/string_array.c:104]: That code is unreachable if `smallest_size` is zero. That's the whole point of moving them into the conditional.

Comment (dirk-thomas) [reply src/string_array.c:104]: I missed that the loop doesn't do any iteration when its zero.

**Comment by cottsay** [comment]
@dirk-thomas, I'm re-requesting review because I added a change to handle some inconsistent behavior due to an assumption based on implementation-specific behavior in malloc.

Here are some docs, which describe what might happen if you malloc/realloc with size 0: http://man7.org/linux/man-pages/man3/malloc.3.html

**Comment by wjwwood** [comment]
@cottsay do you want to try to get this and the other pr's you opened into rcutils before I do a 1.0.0 release? I can try to help make that happen, but I also do not want to delay too long.

**Comment by cottsay** [comment]
> @cottsay do you want to try to get this and the other pr's you opened into rcutils before I do a 1.0.0 release? I can try to help make that happen, but I also do not want to delay too long.

At this point, it would just be "nice to have." I'm not sure I'll have time to actually take advantage of these changes before the release anyway. This one and #252 are a higher priority IMHO - they both improve the behavior of existing functions.

**Comment by cottsay** [comment]
> @cottsay do you want to try to get this and the other pr's you opened into rcutils before I do a 1.0.0 release? I can try to help make that happen, but I also do not want to delay too long.

Since I closed #252, I think you should move ahead with the release without the other two API additions. Thanks, @wjwwood!

**Comment by wjwwood** [comment]
I'm rerunning CI to include more, in case this change in behavior of string array causes a failure downstream:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10902/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6254/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8876/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10793/)

**Comment by wjwwood** [comment]
Oops missed the fact that you merged it. I'll cancel those CI and see what the nightlys bring.

**Pull request merged**